README.md
# XVBA-CLI

## Command Line Interface for XVBA VSCode extension (Under Construction "b" Version)

- This package provides a way to instal VBA packages in Excel with XVBA VSCode extension
- For more details see: <a href="https://www.xvba.dev">www.xvba.dev</a>
- Install packages to your project with command line interface
- <b>Create/Build</b> a VBA package and publish on XVBA repository

## Feedback
- All feedback are welcome
- Mail : alberto.aeraph@gmail.com 
- WhatsApp: +55 31 9 9229 5626 

- Also please write a review, star me on GitHub, and follow me on github
  - Github account: <a href="https://github.com/Aeraphe">Aeraphe</a>

## Install

- Type on terminal and press enter for install the xvba-cli Last Version

```
 npm i @localsmart/xvba-cli@latest --save-dev
```

 <p>
 <img src="https://github.com/aeraphe/images-xvb/raw/master/images/npm_install.gif" alt="Install Packages">
</p>

## List of Commands

### Install Package

- For install new packages use the command 


```
npx xvba install [package]
```

- Example: Instal excel-types package for xvba auto-complete

```
npx xvba install excel-types
```

### List Package

- For list packages use the command

```
npx xvba ls
```

<p>
 <img src="https://github.com/aeraphe/images-xvb/raw/master/images/xvba_ls.gif" alt="List Packages">
</p>

### Uninstall Package 

- For uninstall packages use the command

```
npx xvba uninstall [package]
```



### Create a XVBA package

<p> For Create use the command: (replace the [package] for your package name/folder ) </p>

This command will create a folder and files (on xvba_modules) for you start to create your own package


```
npx xvba create [package]

```

### Build Your Own Package and publish on Xvba Repository

<p> For Build use the command: (replace the [package] for your package name/folder ) </p>

```
npx xvba build [package]

```

### Add VBA files to Your Project

<p> For Build use the command: (replace the [filename] for your package name/filename ) </p>

```
npx xvba add -c [filename] //For classes

npx xvba add -m [filename] //For Modules

```


CHANGELOG.md
# Changelog

## Command Line Interface for XVBA VSCode extension (Under Construction)

- This package provides a way to add VBA packages in Excel with XVBA VSCode extension
- Add third packages to your project with command line interface



## [1.0.0b3] - 2020-09-16
### Added
- New Command Build XVBA Package "xvba build [package]"
- New Command Create XVBA Package "xvba create [package]"

### Fixed
- Refectory all code

## [1.0.0b4] - 2020-09-18
### Added
- New Command Install XVBA Package "xvba install [package]"

### Delete
-- Remove Action Add package from git

### Change
-- Change packages params type array to object

## [1.0.0b6] - 2020-09-22
### Added
- New Command Add VBA class file "xvba  add -c [filename]"  or "xvba  add -c [subfolder/filename]"
- New Command Add VBA Module file "xvba  add -m [filename]" or "xvba  add -m [subfolder/filename]"

## [1.0.0b7] - 2020-09-28
### Fix
- Change file write class example with LF to CRLF


## [1.0.0b11] - 2020-10-27
### Update
- Refactory delete package folder 

actions/create-package.action.js
const fs = require('fs');
const rootPath = process.cwd();
const path = require('path')



module.exports = async function createPackageScaffold(package) {
    try {
        const xvbaModulesDirectory = path.join(rootPath, 'xvba_modules');
        const xvbaModulesPath = path.join(rootPath, 'xvba_modules', package);
        //Developer -> check if the folder has already a folder with the same name
        await handleCheckFolder(xvbaModulesDirectory);
        await handleCheckFolder(xvbaModulesPath);


        await new Promise(
            (resolve, reject) => {
                fs.writeFile(path.join(xvbaModulesPath, 'README.md'), README(package), (error => {
                    error ? reject(error) : resolve(true)
                }))
            }
        )

        await new Promise(
            (resolve, reject) => {
                fs.writeFile(path.join(xvbaModulesPath, 'xvba.package.json'), JSON.stringify(XVBA_PACKAGE(package), undefined, 2), (error => {
                    error ? reject(error) : resolve(true)
                }))
            }
        )
        await new Promise(
            (resolve, reject) => {
                fs.writeFile(path.join(xvbaModulesPath, `${package}_xvba.cls`), INDEX_CLASS(package), (error => {
                    error ? reject(error) : resolve(true)
                }))
            }
        )

        await new Promise(
            (resolve, reject) => {
                fs.writeFile(path.join(xvbaModulesPath, `CHANGELOG.md`), CHANGE_LOG(package), (error => {
                    error ? reject(error) : resolve(true)
                }))
            }
        )

        await new Promise(
            (resolve, reject) => {
                fs.writeFile(path.join(rootPath, 'config.json'), JSON.stringify(CONFIG(package), undefined, 2), (error => {
                    error ? reject(error) : resolve(true)
                }))
            }
        )
    } catch (error) {
        console.log(error)
    }



}

handleCheckFolder = (packagePath) => {
    return new Promise((resolve, reject) => {
        fs.lstat(packagePath, (err, stat) => {
            if (err) {
                try {
                    new Promise(
                        () => {
                            fs.mkdir(packagePath, () => {
                                resolve();
                            })
                        }
                    )
                } catch (error) {
                    reject(error)
                }


            } else {
                resolve(stat.isDirectory())
            }
        })
    })

}


function jsUcfirst(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

/**
 * 
 * @param {String} package 
 */
const INDEX_CLASS = (package) => {
    const name = jsUcfirst(package);

    const classExample = `VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = "${name}"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attribute VB_Exposed = False

'namespace=xvba_modules\\${package}



'Example of Package Create by XVBA-CLI

'Class Param Example
Private pName As String


'Main method example
Private Sub main()

MsgBox "message",,"title"

End Sub


'Example of accessor method Get
Public Property Get Name() As String
 Name = pName
End Property


'Example of accessor method Let
Public Property Let Name(Value as String)
 pName =  Value
End Property

`
    re1 = new RegExp("\x0A", 'g');
    origContent = classExample.replace(re1, "\x0D\x0A");
    return origContent
}




const README = (package) => {
    return `# ${package} (Example of Readme File)
    
## Command ...
 - This package provides a way to ....`
}


const XVBA_PACKAGE = (package) => {
    return {
        "package": package,
        "version": "1.0.0",
        "author": "",
        "email": "",
        "license": "ISC",
        "repository": {
            "type": "git",
            "url": "https://github.com/Aeraphe/xvba-cli.git"
        },
        "description": "Creating amazing XVBA packages",
        "dependencies": {},
        "dev_dependencies": {},
        "homepage": "https://github.com/Aeraphe/xvba-cli"
    }
}


const CHANGE_LOG = (package) => {
    return `# Changelog
## ${package} (Under Construction)

- This package provides a way to add VBA packages in Excel with XVBA VSCode extension
- Add third packages to your project with command line interface
    
## [1.0.0b3] - 2020-09-16
### Added
 - New Command Build XVBA Package "npm xvba build [package]"
 - New Command Create XVBA Package "npm xvba create [package]"

### Fixed
 - Refectory all code
 
 `
}

const CONFIG = (package) => {
    return {
        app_name: "XVBA",
        description: "",
        author: "",
        email: "",
        create_ate: "",
        excel_file: "index.xlsb",
        vba_folder: "vba-files",
        xvba_packages: {},
        xvba_dev_packages: {},
    }
}